File: src/main/java/net/hectus/neobb/turn/legacy_game/other/LTNoteBlock.java
package net.hectus.neobb.turn.legacy_game.other;

import net.hectus.neobb.buff.Buff;
import net.hectus.neobb.player.NeoPlayer;
import net.hectus.neobb.turn.default_game.CounterFilter;
import net.hectus.neobb.turn.default_game.attributes.clazz.RedstoneClazz;
import net.hectus.neobb.turn.default_game.attributes.function.BuffFunction;
import net.hectus.neobb.turn.default_game.attributes.function.CounterFunction;
import net.hectus.neobb.turn.default_game.attributes.function.CounterattackFunction;
import net.hectus.neobb.turn.default_game.block.TGoldBlock;
import net.hectus.neobb.turn.default_game.other.OtherTurn;
import org.bukkit.Instrument;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.block.data.type.NoteBlock;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class LTNoteBlock extends OtherTurn<NoteBlock> implements CounterFunction, BuffFunction, RedstoneClazz {
    public LTNoteBlock(NeoPlayer player) { super(player); }
    public LTNoteBlock(NoteBlock data, Location location, NeoPlayer player) { super(data, location, player); }

    @Override
    public int cost() {
        return 3;
    }

    @Override
    public void apply() {
        new Buff.Luck(Buff.BuffTarget.YOU, luck(data().getInstrument())).apply(player);
    }

    @Override
    public ItemStack item() {
        return new ItemStack(Material.NOTE_BLOCK);
    }

    @Override
    public List<Buff> buffs() {
        return List.of();
    }

    @Override
    public List<CounterFilter> counters() {
        return List.of(CounterFilter.of(t -> t.data() instanceof Block b && b.isSolid() && (!(t instanceof CounterattackFunction) || t instanceof TGoldBlock), "solid"));
    }

    public static int luck(@NotNull Instrument instrument) {
        return switch (instrument) {
            case BELL -> 35;
            case CHIME -> 30;
            case GUITAR, BANJO, COW_BELL -> 20;
            case DIDGERIDOO -> 15;
            case STICKS -> 10;
            case BASS_GUITAR -> 5;
            case BASS_DRUM -> -5;
            default -> -20;
        };
    }
}

File: src/main/kotlin/net/hectus/neobb/modes/turn/legacy_game/other/LTNoteBlock.kt
package net.hectus.neobb.modes.turn.legacy_game.other

import com.marcpg.libpg.storing.Cord
import net.hectus.neobb.buff.Buff
import net.hectus.neobb.buff.Luck
import net.hectus.neobb.modes.turn.default_game.CounterFilter
import net.hectus.neobb.modes.turn.default_game.attribute.clazz.RedstoneClazz
import net.hectus.neobb.modes.turn.default_game.attribute.function.BuffFunction
import net.hectus.neobb.modes.turn.default_game.attribute.function.CounterFunction
import net.hectus.neobb.modes.turn.default_game.attribute.function.CounterattackFunction
import net.hectus.neobb.modes.turn.default_game.block.TGoldBlock
import net.hectus.neobb.modes.turn.default_game.other.OtherTurn
import net.hectus.neobb.player.NeoPlayer
import org.bukkit.Instrument
import org.bukkit.Material
import org.bukkit.block.Block
import org.bukkit.block.data.type.NoteBlock
import org.bukkit.inventory.ItemStack

class LTNoteBlock(data: NoteBlock?, cord: Cord?, player: NeoPlayer?) : OtherTurn<NoteBlock>(data, cord, player), CounterFunction, BuffFunction, RedstoneClazz {
    override val cost: Int = 3

    override fun item(): ItemStack = ItemStack(Material.NOTE_BLOCK)

    override fun apply() {
        Luck(luck(data!!.instrument)).apply(player!!)
    }

    override fun counters(): List<CounterFilter> {
        return listOf(CounterFilter.of("solid") { it.data is Block && it.data.isSolid && (it !is CounterattackFunction || it is TGoldBlock) })
    }

    override fun buffs(): List<Buff<*>> = emptyList()

    fun luck(instrument: Instrument): Int {
        return when (instrument) {
            Instrument.BELL -> 35
            Instrument.CHIME -> 30
            Instrument.GUITAR, Instrument.BANJO, Instrument.COW_BELL -> 20
            Instrument.DIDGERIDOO -> 15
            Instrument.STICKS -> 10
            Instrument.BASS_GUITAR -> 5
            Instrument.BASS_DRUM -> -5
            else -> -20
        }
    }
}
